src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -295,6 +325,93 @@ void ScheduleTask(DateTimeOffset utcNow)
             }
         }
 
+        private readonly List<WeakReference<Stats>?>? _allStats;
+        private readonly Stats? _accumulatedStats;
+        private ThreadLocal<Stats>? _stats;
+
+        private (long, long) Sum()
+        {
+            long hits = 0, misses = 0;
+            lock (_allStats!)
+            {
+                hits += _accumulatedStats!.Hits;
+                misses += _accumulatedStats!.Misses;
+                foreach (WeakReference<Stats>? wr in _allStats!)
+                {
+                    if (wr is not null && wr.TryGetTarget(out Stats? stats))
+                    {
+                        hits += Interlocked.Read(ref stats.Hits);
+                        misses += Interlocked.Read(ref stats.Misses);
+                    }
+                }
+            }
+            return (hits, misses);
+        }
+
+        private void Hit()
+        {
+            if (_options.TrackStatistics)
+            {
+                if (IntPtr.Size == 4)
+                    Interlocked.Increment(ref GetStats().Hits);
+                else
+                    GetStats().Hits++;
+            }
+        }
+
+        private void Miss()
+        {
+            if (_options.TrackStatistics)
+            {
+                if (IntPtr.Size == 4)
+                    Interlocked.Increment(ref GetStats().Misses);
+                else
+                    GetStats().Misses++;
+            }
+        }
+
+        private Stats GetStats() => _stats!.Value!;
+
+        internal sealed class Stats
+        {
+            public long Hits;
+            public long Misses;
+            private readonly MemoryCache _memoryCache;
+            public Stats(MemoryCache memoryCache, bool skipAdd = false)
+            {
+                _memoryCache = memoryCache;
+                if (!skipAdd)
+                    _memoryCache.AddToStats(this);
+            }
+            ~Stats()
+            {
+                _memoryCache.RemoveFromStats(this);
+            }
+        }
+
+        private void RemoveFromStats(Stats current)
+        {
+            lock (_allStats!)
+            {
+                for (int i = 0; i < _allStats!.Count; i++)
+                {
+                    if (_allStats![i] is WeakReference<Stats> wr && wr.TryGetTarget(out Stats? stats) && stats == current)
+                    {
+                        _allStats!.RemoveAt(i);
```

Comment:
I'm assuming that _allStats doesn't get very big? Since the backing storage of the list will never shrink

Reply:
I don't think it would grow very big but I'll investigate to see if we need to set capacity for it

Reply:
_allStats size should grow at most to the number of cores available.

Reply:
ThreadLocal is per-thread rather than per-core, but still agree the size shouldn't be an issue.

Reply:
It should effectively be the max concurrent number of threads that have been alive at the same time (assuming they all access the cache). I'm not concerned about list size, but if we want we could TrimExcess after summing.

Reply:
Added TrimExcess